src/core/util/latent_see.h

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_SRC_CORE_UTIL_LATENT_SEE_H
+#define GRPC_SRC_CORE_UTIL_LATENT_SEE_H
+
+#ifdef GRPC_ENABLE_LATENT_SEE
+#include <chrono>
+#include <cstdint>
+#include <utility>
+#include <vector>
+
+#include "absl/log/log.h"
+
+#include "src/core/lib/gprpp/per_cpu.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_core {
+namespace latent_see {
+
+struct Metadata {
+  const char* file;
+  int line;
+  const char* name;
+};
+
+enum class EventType : uint8_t { kBegin, kEnd, kFlowStart, kFlowEnd, kMark };
+
+class Log {
+ public:
+  static void FlushThreadLog();
+
+  static void Append(const Metadata* metadata, EventType type, uint64_t id) {
+    thread_events_.push_back(
+        Event{metadata, std::chrono::steady_clock::now(), id, type});
+  }
+
+ private:
+  Log() = default;
+
+  static Log& Get() {
+    static Log* log = []() {
+      atexit([] {
+        LOG(INFO) << "Writing latent_see.json in " << get_current_dir_name();
+        FILE* f = fopen("latent_see.json", "w");
+        if (f == nullptr) return;
+        fprintf(f, "%s", log->GenerateJson().c_str());
+        fclose(f);
+      });
+      return new Log();
+    }();
+    return *log;
+  }
+
+  std::string GenerateJson();
+
+  struct Event {
+    const Metadata* metadata;
+    std::chrono::steady_clock::time_point timestamp;
+    uint64_t id;
+    EventType type;
+  };
+  struct RecordedEvent {
+    uint64_t thread_id;
+    uint64_t batch_id;
+    Event event;
+  };
+  std::atomic<uint64_t> next_thread_id_{1};
+  std::atomic<uint64_t> next_batch_id_{1};
+  static thread_local std::vector<Event> thread_events_;
+  static thread_local uint64_t thread_id_;
+  struct Fragment {
+    Mutex mu;
+    std::vector<RecordedEvent> events ABSL_GUARDED_BY(mu);
+  };
+  PerCpu<Fragment> fragments_{PerCpuOptions()};
+};
+
+class Scope {
+ public:
+  explicit Scope(const Metadata* metadata) : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kBegin, 0);
+  }
+  ~Scope() { Log::Append(metadata_, EventType::kEnd, 0); }
+
+  Scope(const Scope&) = delete;
+  Scope& operator=(const Scope&) = delete;
+
+ private:
+  const Metadata* const metadata_;
+};
+
+class ParentScope {
+ public:
+  explicit ParentScope(const Metadata* metadata) : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kBegin, 0);
+  }
+  ~ParentScope() {
+    Log::Append(metadata_, EventType::kEnd, 0);
+    Log::FlushThreadLog();
+  }
+
+  ParentScope(const Scope&) = delete;
+  ParentScope& operator=(const Scope&) = delete;
+
+ private:
+  const Metadata* const metadata_;
+};
+
+class Flow {
+ public:
+  explicit Flow(const Metadata* metadata) : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kFlowStart, id_);
+    Log::FlushThreadLog();
+  }
+  ~Flow() {
+    if (metadata_ != nullptr) {
+      Log::Append(metadata_, EventType::kFlowEnd, id_);
+      Log::FlushThreadLog();
```

Comment:
Should a Flow be required to be in a parent scope instead? Similar to a Scope, which is no-op without having been inside a ParentScope (or followed by a ParentScope, or Flow).

Reply:
I thought I needed this to ensure begin/end ordering, but I don't, so removing it.